game.py

```diff
@@ -411,22 +424,163 @@ def check_ghost_collisions(self):
                 self.reset_positions()
 
     def move_ghost(self, ghost):
-        if ghost.alive == False:
-            # if the ghost is "dead" then we should teleport it back to
-            # its starting location in the house. 
-            ghost.pos[0] = ghost.start_x
-            ghost.pos[1] = ghost.start_y
-            ghost.vulnerable = 0
-            ghost.alive = True
-            ghost.in_house = True
+            # Ghost movement reference: https://gameinternals.com/understanding-pac-man-ghost-behavior
+            # Known deficiencies:
+            #   Pinky and Inky do not chase properly
+            #   
+
+        def remove_backwards(dirs):
+            '''Given a list of directions to move remove the option to reverse direction'''
+            if len(dirs) > 1:
+                if ghost.previous_dir == 'w' and 's' in dirs:
+                    dirs.remove('s')
+                elif ghost.previous_dir == 's' and 'w' in dirs:
+                    dirs.remove('w')
+                elif ghost.previous_dir == 'a' and 'd' in dirs:
+                    dirs.remove('d')
+                elif ghost.previous_dir == 'd' and 'a' in dirs:
+                    dirs.remove('a')
+            
+            return dirs
+        
+        def dir_to_target():
+            '''Find the legal moves that are closest to the current target.'''
+            min_dist = self.map_distance(ghost.pos, ghost.target_pos) + 100
+            dirs = []
+            for delta in [(1, 0, 'd'), (-1, 0, 'a'), (0, 1, 's'), (0, -1, 'w')]:
+                new_pos = (ghost.pos[0] + delta[0], ghost.pos[1] + delta[1])
+                item = self.map[new_pos]
+                print(self.is_blocked(item))
+
+                # Living ghosts can't go back in the house, but dead ones can
+                if ghost.alive:
+                    if (not self.is_blocked(item)) and (not self.is_ghost(item)):
+                        distance = self.map_distance(new_pos, ghost.target_pos)
+                        if distance < min_dist:
+                            min_dist = distance
+                            dirs = [delta[2]]
+                        elif distance == min_dist:
+                            dirs.append(delta[2])
+                else:
+                    if (item == self.DOOR) or (not self.is_blocked and not self.is_ghost(item)):
+                        distance = self.map_distance(new_pos, ghost.target_pos)
+                        if distance < min_dist:
+                            min_dist = distance
+                            dirs = [delta[2]]
+                        elif distance == min_dist:
+                            dirs.append(delta[2])
+            
+            return dirs
+
+        def pick_dir(dirs):
+            '''Given a list of legal and equally good moves, pick one according to "up > left > down"'''
+            choice = dirs[0]
```

Comment:
This assumes that dirs is a non empty list. Will this **always** be true?

Reply:
No, actually that's one of the current bugs I need to fix... From the description of the ghost behavior that I was referring to and the little experience I have playing the arcade game I don't remember ghosts ever sitting still. So I'm pretty sure that assumption _should_ hold, but it currently doesn't.